### PR TITLE
Diagnose "ref" argument in an index operation in strict mode.

### LINKED ISSUE #13062: No error for array index passed by ref
No error reported for `a[ref i]` below:
```
class C
{
    static object F(object[] a, int i)
    {
        return a[ref i];
    }
}
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #13062 

@cston @VSadov Please review
@dotnet/roslyn-compiler FYI

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturns.cs [hunk 1]
```diff
@@ -918,5 +918,32 @@ public static void Main()
...
+            //      The argument-list of an element-access is not allowed to contain ref or out arguments.
+            // However, due to backward compatibility, compiler overlooks this restriction during parsing
+            // and even ignores out/ref modifiers in element access during binding. The "strict" feature
+            // causes the compiler to diagnose this situation.
+            var text = @"
+class C
+{
+    static object F(object[] a, int i)
+    {
+        return a[ref i];
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturns.cs]: Consider testing `a[out i]` as well.

**Comment by cston** [comment]
LGTM

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@cston @AlekseyTs Please re-review. This PR has been revised due to modified compat council outcome: we already allow ref indexers for COM interop (indexed properties) and dynamic indexing, so disallowing it would be a breaking change. We already diagnose ref indexer on index operators when it does not match the underlying API. The only case that needed to change was indexing an array. We now diagnose that as an error.

**Comment by VSadov** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -6181,6 +6164,22 @@ private BoundExpression BindPointerElementAccess(ExpressionSyntax node, BoundExp
...
+        private bool ReportRefOrOutArgument(AnalyzedArguments analyzedArguments, DiagnosticBag diagnostics)
+        {
+            int numArguments = analyzedArguments.Arguments.Count;
+            for (int i = 0; i < numArguments; i++)
+            {
+                RefKind refKind = analyzedArguments.RefKind(i);
+                if (refKind != RefKind.None)
+                {
+                    Error(diagnostics, ErrorCode.ERR_BadArgExtraRef, analyzedArguments.Argument(i).Syntax, i + 1, refKind.ToDisplayString());
+                    return true;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Consider reporting errors for each `ref` argument, not just the first.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 3]
```diff
@@ -6054,7 +6053,7 @@ private BoundExpression BindArrayAccess(ExpressionSyntax node, BoundExpression e
                 }
             }
 
-            return new BoundArrayAccess(node, expr, convertedArguments.AsImmutableOrNull(), arrayType.ElementType);
+            return new BoundArrayAccess(node, expr, convertedArguments.AsImmutableOrNull(), arrayType.ElementType, hasErrors);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: We aren't setting `hasErrors` for errors converting array index arguments above. Is there a reason to set `hasErrors` for `ref` arguments errors only?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 4]
```diff
@@ -6181,6 +6164,22 @@ private BoundExpression BindPointerElementAccess(ExpressionSyntax node, BoundExp
             return new BoundPointerElementAccess(node, expr, index, CheckOverflowAtRuntime, pointedAtType, hasErrors);
         }
 
+        private bool ReportRefOrOutArgument(AnalyzedArguments analyzedArguments, DiagnosticBag diagnostics)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: `static`

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs [hunk 5]
```diff
@@ -918,5 +918,34 @@ public static void Main()
...
+        {
+            var text = @"
+class C
+{
+    void F(object[] a, int i)
+    {
+        int j;
+        j = a[ref i];    // error 1
+        j = a[out i];    // error 2
+        j = this[ref i]; // error 3
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Consider testing other indices: `a[0, ref i]`

Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: Consider testing other indices: `a[out i, ref j]`

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: This is the existing behavior for the contexts in which this is diagnosed.

---
In reply to: [75064582](https://github.com/dotnet/roslyn/pull/13091#discussion_r75064582) [](ancestors = 75064582)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 3 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Not really. The reported diagnostic should prevent downstream processing anyway. I'm just preserving existing behavior while adding this error.

---
In reply to: [75064678](https://github.com/dotnet/roslyn/pull/13091#discussion_r75064678) [](ancestors = 75064678)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: yes

---
In reply to: [75064780](https://github.com/dotnet/roslyn/pull/13091#discussion_r75064780) [](ancestors = 75064780)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs (hunk 5 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs]: yes

---
In reply to: [75065315](https://github.com/dotnet/roslyn/pull/13091#discussion_r75065315) [](ancestors = 75065315)

**Comment by cston** [comment]
LGTM

**Pull request closed without merging**